File: code/ch10/10.2.4-house-rob.java
class Solution {

    public int rob(int[] nums) {
        if(nums.length <= 0) return 0;

        int prev = 0, curr = 0;

        for (int i = nums.length-1; i >= 0; i--) {
            int tmp = curr;
            curr = Math.max(curr, nums[i] + prev);
            prev = tmp;
        }
        return curr;
    }
}

File: code/ch10/10.2.4-house-rob.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() <= 0) return 0;

        int prev, curr;

        for(int i = nums.size()-1; i >= 0; i--) {
            int tmp = curr;
            curr = max(curr, nums[i] + prev);
            prev = tmp;
        }
        return curr;
    }
};